src/factotum_server/mod.rs

```
@@ -0,0 +1,242 @@
+        }
+    };
+    tx.send(result).expect("Server status channel receiver has been deallocated");
+}
+
+fn is_queue_full(query: Query<bool>, requests_queue: &mut VecDeque<JobRequest>, max_jobs: usize) {
+    let tx = query.status_tx;
+    let is_full = requests_queue.len() >= max_jobs;
+    tx.send(is_full).expect("Queue query channel receiver has been deallocated");
+}
+
+fn new_job_request<T: Persistence>(requests_channel: Sender<Dispatch>, requests_queue: &mut VecDeque<JobRequest>, primary_pool: ThreadPool, persistence: T, request: JobRequest) {
+    debug!("ADDING NEW JOB jobId:[{}]", request.job_id);
+    requests_queue.push_back(request.clone());
+    // Create entry in persistence storage
+    persist_entry(&persistence, request.job_id.clone(), request.clone(), JobState::Queued);
+    // Check queue size - return error if limit exceeded (not important right now)
+    if primary_pool.active_count() < primary_pool.max_count() {
+        requests_channel.send(Dispatch::ProcessRequest).expect("Job requests channel receiver has been deallocated");
+    } else {
+        info!("No threads available - waiting for a job to complete.")
+    }
+}
+
+fn process_job_request<T: 'static + Persistence + Send>(requests_channel: Sender<Dispatch>, requests_queue: &mut VecDeque<JobRequest>, primary_pool: ThreadPool, persistence: T, command_store: CommandStore) {
+    debug!("QUEUE SIZE = {}", requests_queue.len());
+    match requests_queue.pop_front() {
+        Some(request) => {
+            primary_pool.execute(move || {
+                debug!("PROCESSING JOB REQ jobId:[{}]", request.job_id);
+                // Update status in persistence storage
+                persist_entry(&persistence, request.job_id.clone(), request.clone(), JobState::Working);
+                let cmd_path = match command_store.get_command(::FACTOTUM) {
+                    Ok(path) => path,
+                    Err(e) => {
+                        error!("{}", e);
```

This function could return `Err` too - if it isn't required errors can be ignored